File: tests/main/commands.test.ts
import { describe, it, expect, beforeEach } from 'vitest'
import { runCommand } from '../../src/main/irc/commands'
import { serializeMessage } from '../../src/main/irc/serializer'
import { readFileSync } from 'fs'
import { join } from 'path'

interface Call {
  method: string
  args: unknown[]
}

let calls: Call[] = []

function record(method: string) {
  return (...args: unknown[]) => {
    calls.push({ method, args })
  }
}

/** Stands in for IRCClient — records what a command asked the client to do */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const client = {
  action: record('action'),
  join: record('join'),
  part: record('part'),
  setNick: record('setNick'),
  say: record('say'),
  notice: record('notice'),
  whois: record('whois'),
  setTopic: record('setTopic'),
  mode: record('mode'),
  kick: record('kick'),
  disconnect: record('disconnect'),
  connection: {
    send: record('send'),
    sendRaw: record('sendRaw')
  },
  // What the server said it will take. Empty by default — a server that states
  // nothing places no limit, which is the case most of these run under.
  state: { isupport: {} as Record<string, string | true> }
} as any

const run = (text: string, target = '#chan') => runCommand(client, target, text)

beforeEach(() => {
  calls = []
  // A server that states nothing places no limit, which is what most of these
  // run under; the cases that care set their own.
  client.state.isupport = {}
})

describe('slash commands', () => {
  it('leaves ordinary text alone', () => {
    expect(run('hello there').handled).toBe(false)
    expect(calls).toEqual([])
  })

  it('sends a literal message for a doubled slash', () => {
    const result = run('//not a command')
    expect(result.handled).toBe(false)
    expect(result.message).toBe('/not a command')
  })

  it('never sends an unknown command as a message', () => {
    const result = run('/definitelynotacommand some args')
    expect(result.handled).toBe(true)
    expect(result.message).toBeUndefined()
    expect(result.error).toMatch(/unknown command/i)
    expect(calls).toEqual([])
  })

  it('keeps /msg out of the channel', () => {
    // The bug this guards: "/msg NickServ IDENTIFY hunter2" as a public message
    const result = run('/msg NickServ IDENTIFY hunter2')
    expect(result.handled).toBe(true)
    expect(result.error).toBeUndefined()
    expect(calls).toEqual([{ method: 'say', args: ['NickServ', 'IDENTIFY hunter2'] }])
  })

  it('runs /me as an action', () => {
    run('/me waves hello')
    expect(calls).toEqual([{ method: 'action', args: ['#chan', 'waves hello'] }])
  })

  it('joins a channel, adding the # when left off', () => {
    run('/join #foo')
    run('/join bar')
    run('/join #baz secretkey')
    expect(calls).toEqual([
      { method: 'join', args: ['#foo', undefined] },
      { method: 'join', args: ['#bar', undefined] },
      { method: 'join', args: ['#baz', 'secretkey'] }
    ])
  })

  it('parts the current channel by default, or a named one', () => {
    run('/part')
    run('/part later everyone')
    run('/part #other bye')
    expect(calls).toEqual([
      { method: 'part', args: ['#chan', undefined] },
      { method: 'part', args: ['#chan', 'later everyone'] },
      { method: 'part', args: ['#other', 'bye'] }
    ])
  })

  it('sets and queries the topic', () => {
    run('/topic')
    run('/topic a brand new topic')
    expect(calls).toEqual([
      { method: 'send', args: ['TOPIC', '#chan'] },
      { method: 'setTopic', args: ['#chan', 'a brand new topic'] }
    ])
  })

  it('applies /mode to the current channel when given only flags', () => {
    run('/mode +o someone')
    expect(calls).toEqual([{ method: 'mode', args: ['#chan', '+o', 'someone'] }])
  })

  it('applies /mode to an explicit target', () => {
    run('/mode #other +m')
    expect(calls).toEqual([{ method: 'mode', args: ['#other', '+m'] }])
  })

  it('kicks with an optional reason', () => {
    run('/kick troublemaker being rude')
    expect(calls).toEqual([{ method: 'kick', args: ['#chan', 'troublemaker', 'being rude'] }])
  })

  it('refuses channel-only commands in a DM', () => {
    expect(run('/kick someone', 'alice').error).toMatch(/only works in a channel/)
    expect(run('/topic new', 'alice').error).toMatch(/only works in a channel/)
    expect(calls).toEqual([])
  })

  it('sets and clears away', () => {
    run('/away making coffee')
    run('/away')
    run('/back')
    expect(calls).toEqual([
      { method: 'send', args: ['AWAY', 'making coffee'] },
      { method: 'send', args: ['AWAY'] },
      { method: 'send', args: ['AWAY'] }
    ])
  })

  it('reports usage instead of acting on an incomplete command', () => {
    expect(run('/join').error).toMatch(/usage/i)
    expect(run('/nick').error).toMatch(/usage/i)
    expect(run('/msg someone').error).toMatch(/nothing to send/i)
    expect(calls).toEqual([])
  })

  it('passes /raw through untouched', () => {
    run('/raw PRIVMSG #chan :hi')
    expect(calls).toEqual([{ method: 'sendRaw', args: ['PRIVMSG #chan :hi'] }])
  })
})

/**
 * Limits the server states but does not enforce with an error.
 *
 * `TOPICLEN` and its neighbours are not refusals: go over one and the server
 * accepts the command and quietly keeps the first N bytes, so the user finds
 * out later, if at all.
 */
describe('text the server would silently cut', () => {
  beforeEach(() => {
    client.state.isupport = { TOPICLEN: '20', AWAYLEN: '10', KICKLEN: '8' }
  })

  it('refuses a topic too long for this network, and says by how much', () => {
    const result = run(`/topic ${'t'.repeat(40)}`)

    expect(result.error).toContain('20')
    expect(result.error).toContain('40')
    expect(calls.find((c) => c.method === 'setTopic')).toBeUndefined()
  })

  it('lets a topic that fits through', () => {
    run('/topic a short one')
    expect(calls).toContainEqual({ method: 'setTopic', args: ['#chan', 'a short one'] })
  })

  it('refuses an away message too long for this network', () => {
    expect(run(`/away ${'a'.repeat(30)}`).error).toContain('10')
    expect(calls.find((c) => c.method === 'send')).toBeUndefined()
  })

  /** Clearing it is not a message, so there is nothing to be too long */
  it('still lets you come back', () => {
    run('/away')
    expect(calls).toContainEqual({ method: 'send', args: ['AWAY'] })
  })

  it('refuses a kick reason too long for this network', () => {
    expect(run(`/kick robin ${'r'.repeat(20)}`).error).toContain('8')
    expect(calls.find((c) => c.method === 'kick')).toBeUndefined()
  })

  it('kicks with no reason at all regardless', () => {
    run('/kick robin')
    expect(calls).toContainEqual({ method: 'kick', args: ['#chan', 'robin', undefined] })
  })

  /** Bytes, not characters — which is what the server counts */
  it('counts a limit in bytes rather than characters', () => {
    client.state.isupport = { TOPICLEN: '10' }

    // Five characters, fifteen bytes
    expect(run('/topic 日本語です').error).toContain('15')
  })

  it('places no limit when the server states none', () => {
    client.state.isupport = {}
    run(`/topic ${'t'.repeat(500)}`)

    expect(calls.find((c) => c.method === 'setTopic')).toBeDefined()
  })
})

/**
 * The new commands, as lines on the wire.
 *
 * Compared against `tests/fixtures/commands.json`, which the Android
 * `CommandsTest` reads too — so `/op` on the phone and `/op` on the desktop
 * produce the same bytes, or one of the two tests goes red.
 *
 * Lines rather than method calls, because the two clients reach the socket
 * through differently-shaped objects and the only thing they genuinely share
 * is what comes out of it.
 */
describe('the commands every client has', () => {
  const corpus = JSON.parse(
    readFileSync(join(__dirname, '../fixtures/commands.json'), 'utf8')
  ) as {
    cases: {
      name: string
      input: string
      target: string
      nick: string
      isupport: Record<string, string>
      lines?: string[]
      error?: string
    }[]
  }

  for (const c of corpus.cases) {
    it(c.name, () => {
      const lines: string[] = []
      const wire = {
        connection: {
          send: (...params: string[]) => lines.push(serialise(params)),
          sendRaw: (line: string) => lines.push(line)
        },
        state: {
          isupport: c.isupport as Record<string, string | true>,
          nick: c.nick,
          casemap: (value: string) => value.toLowerCase(),
          channels: new Map()
        },
        // The convenience methods IRCClient offers, as the lines they send
        mode: (target: string, mode: string, ...params: string[]) =>
          lines.push(serialise(['MODE', target, mode, ...params])),
        kick: (channel: string, nick: string, reason?: string) =>
          lines.push(serialise(['KICK', channel, nick, reason ?? nick])),
        join: (channel: string) => lines.push(serialise(['JOIN', channel])),
        part: (channel: string) => lines.push(serialise(['PART', channel]))
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } as any

      const result = runCommand(wire, c.target, c.input)
      expect(result.handled).toBe(true)

      if (c.error) {
        expect(result.error).toBe(c.error)
        expect(lines).toEqual([])
      } else {
        expect(result.error).toBeUndefined()
        expect(lines).toEqual(c.lines)
      }
    })
  }

  /**
   * The one that would be a memorable bug. Kicking before banning leaves a
   * window in which they can rejoin, which is the single thing kickban exists
   * to prevent.
   */
  it('bans before it kicks, never after', () => {
    const order = corpus.cases.find((c) => c.name === '/kickban bans before it kicks')!
    expect(order.lines![0].startsWith('MODE')).toBe(true)
    expect(order.lines![1].startsWith('KICK')).toBe(true)
  })
})

/**
 * The real serializer, not a second copy of its rules.
 *
 * Which parameters go in the trailing form is exactly the sort of thing the
 * two clients could disagree about, so the comparison has to run through the
 * code that actually decides it.
 */
function serialise(params: string[]): string {
  const [command, ...rest] = params
  return serializeMessage({ command, params: rest })
}

File: android/app/src/test/java/org/switchboard/android/CommandsTest.kt
package org.switchboard.android

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.switchboard.android.irc.Commands
import org.switchboard.android.irc.ConnectionState
import org.switchboard.android.irc.Irc
import org.switchboard.android.irc.IrcCommandTarget
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Slash commands on the phone.
 *
 * These exist because they did not: holding its own connection, the phone sent
 * everything typed straight to the channel, so `/msg NickServ IDENTIFY hunter2`
 * published the password to the room. The first test is that one.
 *
 * The rest check the same answers `src/main/irc/commands.ts` gives, because two
 * clients that disagree about what `/part` means are two clients.
 */
class CommandsTest {

    /** A connection that records instead of connecting */
    private class Target(override val state: ConnectionState) : IrcCommandTarget {
        val calls = mutableListOf<String>()

        override fun send(command: String, vararg params: String) {
            calls.add("send $command ${params.joinToString(" ")}".trim())
        }
        override fun sendRaw(line: String) { calls.add("raw $line") }
        override fun say(target: String, text: String) { calls.add("say $target :$text") }
        override fun action(target: String, text: String) { calls.add("action $target :$text") }
        override fun notice(target: String, text: String) { calls.add("notice $target :$text") }
        override fun join(channel: String, key: String?) {
            calls.add("join $channel${key?.let { " $it" } ?: ""}")
        }
        override fun part(channel: String, reason: String?) {
            calls.add("part $channel${reason?.let { " :$it" } ?: ""}")
        }
        override fun setTopic(channel: String, topic: String) { calls.add("topic $channel :$topic") }
        override fun setNick(nick: String) { calls.add("nick $nick") }
        override fun whois(nick: String) { calls.add("whois $nick") }
        override fun setAway(message: String?) { calls.add("away ${message ?: ""}".trim()) }
        override fun setMode(target: String, mode: String, vararg args: String) {
            calls.add("mode $target $mode ${args.joinToString(" ")}".trim())
        }
        override fun kick(channel: String, nick: String, reason: String?) {
            calls.add("kick $channel $nick${reason?.let { " :$it" } ?: ""}")
        }
        override fun invite(nick: String, channel: String) { calls.add("invite $nick $channel") }
        override fun stop(quitMessage: String) { calls.add("quit :$quitMessage") }
    }

    private lateinit var target: Target

    @Before
    fun setUp() {
        val state = ConnectionState("srv")
        state.reset("kara")
        target = Target(state)
    }

    private fun run(text: String, on: String = "#chan") = Commands.run(target, on, text)

    // ── the reason this file exists ──────────────────────────────────

    @Test
    fun `a password sent to NickServ does not reach the channel`() {
        val result = run("/msg NickServ IDENTIFY hunter2")

        assertTrue(result.handled)
        assertNull(result.error)
        assertEquals(listOf("say NickServ :IDENTIFY hunter2"), target.calls)
    }

    @Test
    fun `an unknown command is refused rather than said out loud`() {
        val result = run("/idenify hunter2")

        assertTrue(result.handled)
        assertEquals("Unknown command: /idenify", result.error)
        assertTrue(target.calls.isEmpty())
    }

    @Test
    fun `ordinary text is not a command`() {
        val result = run("hello everyone")

        assertFalse(result.handled)
        assertNull(result.message)
        assertTrue(target.calls.isEmpty())
    }

    @Test
    fun `a doubled slash sends one literal slash`() {
        val result = run("//afk for a bit")

        assertFalse(result.handled)
        assertEquals("/afk for a bit", result.message)
    }

    // ── the everyday ones ────────────────────────────────────────────

    @Test
    fun `join adds the hash when it was left off`() {
        run("/join lobby")
        assertEquals(listOf("join #lobby"), target.calls)
    }

    @Test
    fun `join passes a channel key through`() {
        run("/j #secret sesame")
        assertEquals(listOf("join #secret sesame"), target.calls)
    }

    @Test
    fun `part with no argument leaves the channel you are in`() {
        run("/part", on = "#lobby")
        assertEquals(listOf("part #lobby"), target.calls)
    }

    @Test
    fun `part with a reason keeps the reason whole`() {
        run("/part bye for now", on = "#lobby")
        assertEquals(listOf("part #lobby :bye for now"), target.calls)
    }

    @Test
    fun `part naming another channel leaves that one instead`() {
        run("/part #elsewhere gone", on = "#lobby")
        assertEquals(listOf("part #elsewhere :gone"), target.calls)
    }

    @Test
    fun `me sends an action`() {
        run("/me waves")
        assertEquals(listOf("action #chan :waves"), target.calls)
    }

    @Test
    fun `me with nothing to do says so`() {
        assertEquals("/me needs something to do", run("/me").error)
        assertTrue(target.calls.isEmpty())
    }

    @Test
    fun `topic only works in a channel`() {
        val result = run("/topic hello", on = "robin")

        assertEquals("/topic only works in a channel", result.error)
        assertTrue(target.calls.isEmpty())
    }

    @Test
    fun `topic with no text asks what the topic is`() {
        run("/topic", on = "#lobby")
        assertEquals(listOf("send TOPIC #lobby"), target.calls)
    }

    /**
     * TOPICLEN is not a refusal on the wire — the server takes the command and
     * silently keeps the first N bytes. Saying so now is the only chance the
     * user gets to choose which half survives.
     */
    @Test
    fun `a topic longer than the network allows is refused with the numbers`() {
        target.state.isupport["TOPICLEN"] = "10"

        val result = run("/topic a topic considerably longer than ten", on = "#lobby")

        assertEquals(
            "This network allows 10 characters in a topic and yours is 36.",
            result.error
        )
        assertTrue(target.calls.isEmpty())
    }

    @Test
    fun `a topic is measured in bytes, the way the server measures it`() {
        target.state.isupport["TOPICLEN"] = "8"

        // Three characters, nine bytes: a client counting characters would let
        // this through and watch the server cut it.
        assertTrue(run("/topic 日本語", on = "#lobby").error!!.endsWith("yours is 9."))
    }

    @Test
    fun `mode on the current channel is recognised by its shape`() {
        run("/mode +o robin", on = "#lobby")
        assertEquals(listOf("mode #lobby +o robin"), target.calls)
    }

    @Test
    fun `mode naming a target uses it`() {
        run("/mode #other +m", on = "#lobby")
        assertEquals(listOf("mode #other +m"), target.calls)
    }

    @Test
    fun `mode with a single argument asks for that target's modes`() {
        run("/mode #other", on = "#lobby")
        assertEquals(listOf("send MODE #other"), target.calls)
    }

    @Test
    fun `kick keeps the reason whole and only works in a channel`() {
        run("/kick robin being tiresome", on = "#lobby")
        assertEquals(listOf("kick #lobby robin :being tiresome"), target.calls)

        target.calls.clear()
        assertEquals("/kick only works in a channel", run("/kick robin", on = "robin").error)
        assertTrue(target.calls.isEmpty())
    }

    @Test
    fun `away with no message comes back`() {
        run("/away")
        run("/back")
        assertEquals(listOf("away", "away"), target.calls)
    }

    @Test
    fun `an away message longer than AWAYLEN is refused`() {
        target.state.isupport["AWAYLEN"] = "5"
        assertTrue(run("/away out for lunch").error!!.startsWith("This network allows 5"))
    }

    @Test
    fun `invite defaults to the channel you are in`() {
        run("/invite robin", on = "#lobby")
        assertEquals(listOf("invite robin #lobby"), target.calls)
    }

    @Test
    fun `quit carries the message you gave it`() {
        run("/quit see you")
        assertEquals(listOf("quit :see you"), target.calls)
    }

    @Test
    fun `raw sends the line as typed`() {
        run("/raw WHO #lobby %tcuhnfar")
        assertEquals(listOf("raw WHO #lobby %tcuhnfar"), target.calls)
    }

    @Test
    fun `nick and whois take one argument and complain without it`() {
        run("/nick karaz")
        run("/whois robin")
        assertEquals(listOf("nick karaz", "whois robin"), target.calls)

        target.calls.clear()
        assertEquals("Usage: /nick <nickname>", run("/nick").error)
        assertEquals("Usage: /whois <nick>", run("/whois").error)
    }

    @Test
    fun `msg without a body says what is missing`() {
        assertEquals("Nothing to send to robin", run("/msg robin").error)
        assertTrue(target.calls.isEmpty())
    }

    @Test
    fun `commands are recognised whatever case they are typed in`() {
        run("/JOIN #lobby")
        assertEquals(listOf("join #lobby"), target.calls)
    }

    /**
     * Nothing we send may carry a newline.
     *
     * `/quit <message>` is the one command whose text reaches the socket
     * without passing the send queue — a QUIT is written immediately so that
     * leaving is not held up behind whatever else was waiting — which is
     * exactly why the check belongs at the socket rather than on the queue.
     */
    @Test
    fun `a newline in what was typed cannot become a second command`() {
        assertEquals(
            "QUIT :byeJOIN #elsewhere",
            Irc.oneLine(Irc.serialise("QUIT", listOf("bye\r\nJOIN #elsewhere")))
        )
        assertEquals(
            "TOPIC #chan :helloJOIN #elsewhere",
            Irc.oneLine(Irc.serialise("TOPIC", listOf("#chan", "hello\nJOIN #elsewhere")))
        )
        assertEquals("nothing to strip", Irc.oneLine("nothing to strip"))
    }

    // ── the same lines the desktop sends ─────────────────────────────

    /**
     * A connection that records the *lines*, not the calls.
     *
     * The desktop reaches the socket through a differently shaped object, so
     * the only thing the two genuinely share is what comes out of it. Anything
     * compared above the wire is comparing two APIs, not two clients.
     */
    private class Wire(override val state: ConnectionState) : IrcCommandTarget {
        val lines = mutableListOf<String>()

        private fun line(command: String, vararg params: String) {
            lines.add(Irc.serialise(command, params.toList()))
        }

        override fun send(command: String, vararg params: String) = line(command, *params)
        override fun sendRaw(line: String) { lines.add(line) }
        override fun say(target: String, text: String) = line("PRIVMSG", target, text)
        override fun action(target: String, text: String) =
            line("PRIVMSG", target, "\u0001ACTION $text\u0001")
        override fun notice(target: String, text: String) = line("NOTICE", target, text)
        override fun join(channel: String, key: String?) =
            if (key == null) line("JOIN", channel) else line("JOIN", channel, key)
        override fun part(channel: String, reason: String?) =
            if (reason == null) line("PART", channel) else line("PART", channel, reason)
        override fun setTopic(channel: String, topic: String) = line("TOPIC", channel, topic)
        override fun setNick(nick: String) = line("NICK", nick)
        override fun whois(nick: String) = line("WHOIS", nick)
        override fun setAway(message: String?) =
            if (message == null) line("AWAY") else line("AWAY", message)
        override fun setMode(target: String, mode: String, vararg args: String) =
            line("MODE", target, mode, *args)
        override fun kick(channel: String, nick: String, reason: String?) =
            line("KICK", channel, nick, reason ?: nick)
        override fun invite(nick: String, channel: String) = line("INVITE", nick, channel)
        override fun stop(quitMessage: String) = line("QUIT", quitMessage)
    }

    /**
     * Every case in `tests/fixtures/commands.json`, which the desktop's
     * `commands.test.ts` reads too. `/op` on the phone and `/op` on the desktop
     * produce the same bytes, or one of the two goes red.
     */
    @Test
    fun `sends the same lines the desktop sends`() {
        val corpus = Json.parseToJsonElement(
            java.io.File(
                System.getProperty("switchboard.fixtures")!!,
                "commands.json"
            ).readText()
        ).jsonObject

        for (case in corpus["cases"]!!.jsonArray) {
            val c = case.jsonObject
            val name = c["name"]!!.jsonPrimitive.content

            val state = ConnectionState("srv")
            state.reset(c["nick"]!!.jsonPrimitive.content)
            for ((key, value) in c["isupport"]!!.jsonObject) {
                state.isupport[key] = value.jsonPrimitive.content
            }

            val wire = Wire(state)
            val result = Commands.run(wire, c["target"]!!.jsonPrimitive.content, c["input"]!!.jsonPrimitive.content)

            assertTrue(name, result.handled)
            val error = c["error"]?.jsonPrimitive?.content
            if (error != null) {
                assertEquals(name, error, result.error)
                assertEquals(name, emptyList<String>(), wire.lines)
            } else {
                assertNull(name, result.error)
                assertEquals(
                    name,
                    c["lines"]!!.jsonArray.map { it.jsonPrimitive.content },
                    wire.lines
                )
            }
        }
    }
}
